### PR TITLE
Add type annotations to xml.py

### PR DESCRIPTION
**Pull request by trim21**

### CODE REVIEW

**Review by balamurugana — CHANGES_REQUESTED** [review]
Duck typing in this PR is incomplete. Please fix accordingly.

**Comment by trim21** [comment]
> Duck typing in this PR is incomplete. Please fix accordingly.

Sorry, which class do you mean? `IFromXML` and `IToXML` are complete?

**Comment by trim21** [comment]
> > > Duck typing in this PR is incomplete. Please fix accordingly.
> > 
> > 
> > Sorry, which class do you mean? `IFromXML` and `IToXML` are incomplete?
> 
> Code in xml.py

done

File: minio/xml.py [hunk 1]
```diff
@@ -80,20 +92,43 @@ def findtext(element, name, strict=False):
...
+    with io.BytesIO() as data:
+        ET.ElementTree(element).write(
+            data,
+            encoding=None,
+            xml_declaration=False,
+        )
+        return data.getvalue()
+
+
+class IToXML(Protocol):
```
Comment (balamurugana) [inline minio/xml.py:125]: How about using abstract class instead of `Protocol` inheritance?

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: minio/xml.py (hunk 1 above)
Comment (trim21) [reply minio/xml.py:125]: That would involve many other files

Comment (balamurugana) [reply minio/xml.py:125]: I couldn't able to understand why it involves other files.

Comment (trim21) [reply minio/xml.py:125]: abstract class doesn't work like `Protocol` in typing, abstract class is expected to be inherited as base class, but in typing `Protocol` mean duck typing.

This mean, when a class support `toxml(...)`, this class need to use `IToXMl` as base class, which dould involve other files.

Comment (trim21) [reply minio/xml.py:125]: example:
```python
import abc
from abc import ABC


class ABCToXML(ABC):

    @abc.abstractmethod
    def toxml(self, element: str | None) -> str:
        ...


class ImplCorrect(ABCToXML):
    """"""

    def toxml(self, element: str | None) -> str:
        return ""


class ConsiderNotImplemented:
    """"""

    def toxml(self, element: str | None) -> str:
        return ""


_1: ABCToXML = ImplCorrect()
_2: ABCToXML = ConsiderNotImplemented() # type error here
```

File: minio/xml.py [hunk 2]
```diff
@@ -80,20 +92,43 @@ def findtext(element, name, strict=False):
     return element.text or ""
 
 
-def unmarshal(cls, xmlstring):
+K = TypeVar("K")
```
Comment (balamurugana) [inline minio/xml.py:95]: Have meaningful name here

File: minio/xml.py [hunk 3]
```diff
@@ -80,20 +92,43 @@ def findtext(element, name, strict=False):
...
+
+
+class IFromXML(Protocol):
```
Comment (balamurugana) [inline minio/xml.py:98]: Rename this class as `FromXMLType` or something meaningful

File: minio/xml.py [hunk 4]
```diff
@@ -80,20 +92,43 @@ def findtext(element, name, strict=False):
...
+    """typing stub for class with `fromxml` method"""
+
+    @classmethod
+    def fromxml(cls: Type[K], element: ET.Element) -> K:
+        """Create python object with values from XML element."""
+
+
+T = TypeVar("T", bound=IFromXML)
```
Comment (balamurugana) [inline minio/xml.py:106]: Have a meaningful name here

File: minio/xml.py (hunk 1 above)
Comment (balamurugana) [inline minio/xml.py:125]: Have a meaningful name here

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: minio/xml.py (hunk 2 above)
Comment (trim21) [reply minio/xml.py:95]: this is generic type parameter, it's custom to be named as `T` `K` `J`。

File: minio/xml.py [hunk 5]
```diff
@@ -80,20 +92,43 @@ def findtext(element, name, strict=False):
...
 
-def getbytes(element):
+def getbytes(element: ET.Element) -> bytes:
     """Convert ElementTree.Element to bytes."""
-    data = io.BytesIO()
-    ET.ElementTree(element).write(
-        data, encoding=None, xml_declaration=False,
-    )
-    return data.getvalue()
+    with io.BytesIO() as data:
```
Comment (balamurugana) [inline minio/xml.py:116]: Why do we need to use `with` here?

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: minio/xml.py (hunk 5 above)
Comment (trim21) [reply minio/xml.py:116]: close file and discard the buffer.

Comment (trim21) [reply minio/xml.py:116]: https://docs.python.org/3/library/io.html#io.BytesIO

Comment (trim21) [reply minio/xml.py:116]: This is not necessary, gc will handle them if we don't use `with`

Comment (balamurugana) [reply minio/xml.py:116]: Yep. For this function, it is unnecessary.

**Review by balamurugana — APPROVED** [review]

File: minio/xml.py [hunk 6]
```diff
@@ -16,28 +16,36 @@
...
+) -> ET.Element:
     """Create ElementTree.Element with tag and namespace."""
-    return ET.Element(tag, {'xmlns': namespace} if namespace else {})
+    return ET.Element(tag, {"xmlns": namespace} if namespace else {})
 
 
-def SubElement(parent, tag, text=None):  # pylint: disable=invalid-name
+def SubElement(  # pylint: disable=invalid-name
+    parent: ET.Element, tag: str, text: str | None = None
+) -> ET.Element:
```
Comment (balamurugana) [inline minio/xml.py:40]: It returns `ET.SubElement`

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: minio/xml.py (hunk 6 above)
Comment (trim21) [inline minio/xml.py:40]:
```suggestion
) -> ET.SubElement:
```

**Review by balamurugana — APPROVED** [review]

**Pull request merged**